### PR TITLE
[requirements] Split requirements into different tiers [build_base]

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This PR splits the requirements files into different tiers:

- `air/core-requirements.txt` - these are the most important requirements (tier 1)
- `air/[library]-requirements.txt` - these are pinned dependencies for libraries (tier 2+3)
- `air/[library]-test-requirements.txt` - these are pinned test-dependencies for libraries (tier 4)

(tier 2 and 3 may get merged into one tier, tbd).

This does not change any packages and should not affect installed packages in docker or CI.

The naming format was chosen to be compatible with dependabot.

In a follow-up PR, we will compile the tier 1-3 dependencies to ensure all CI runners and Docker images run on the same dependency versions.

This PR will split existing requirement files. In total, there will be 4 more such files (`rllib-test-requirements.txt`, `tune-test-requirements.txt`, `train-test-requirements.txt`, `docker/ray-docker-requirements.txt`

## Background

We currently ship different subdependencies in different environments. The packages installed in CI differ between jobs, and the packages in the docker images differ between python versions and image type (ray vs ray-ml).

The reason for that is in those environments only a subset of dependencies are installed. Generally the more dependencies are installed in one go, the more restrictive the subdependencies are and the older the shipped versions. This can lead to a situation where we ship packages in our docker images that are not tested on CI and can actually fail.

The plan going forward is to pip-compile the full list of dependencies into a constraints file. This constraints file is then used in subsequent jobs to install the packages. By utilizing the constraints file we can ensure that the package versions match between CI and docker containers, but we don't have to install all packages in all environments.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/requirements/air/core-requirements.txt
Comment (matthewdeng) [inline python/requirements/air/core-requirements.txt:1]: It probably doesn't make sense to have all these in an `air` directory, I wonder if we can just put these directly in `requirements` instead.

@pcmoritz PTAL and leave any suggestions on how to organize this.

**Review by can-anyscale — APPROVED** [review]

**Comment by aslonnie** [comment]
I feel that this setup is going to be quite hard to maintain over time.

like if I add a dependency, or require pinning a transient dependency, how do I know / keep track which file I should update?

I think in the long run:

- ray needs to be properly split into sub packages / libraries. we can still release everything in one, but internally, we need to have clearer module boundaries and layers that is enforced by build and test procedures.
- based on the package structure, we need to compile a coherently resolved dependency graph for all the dependencies, for each python+platform combination.
- and then, based on this big dependency graph, we can extract the smaller sub graphs for each module / component, and use the sub graph to build the containers or run the tests.

if we use bazel to compile the dependencies, this is basically what bazel will do. sadly, bazel does not have ver great support for multi-python+multi-platform, or at least it will take a lot of efforts.

that said, the dependency graph compilation for third-party dependencies is basically done by pip, which means we can do it in our own way if we want to.

maybe the challenging part is actually "properly splitting ray into sub components".

**Comment by krfricke** [comment]
> I feel that this setup is going to be quite hard to maintain over time.

It's what we're doing right now, but the renaming actually enables e.g. dependabot to make updates easier in the future.

> 
> like if I add a dependency, or require pinning a transient dependency, how do I know / keep track which file I should update?

Again no real change here, usually we pin these where the core dependency is defined.

> 
> I think in the long run:
> 
>     * ray needs to be properly split into sub packages / libraries. we can still release everything in one, but internally, we need to have clearer module boundaries and layers that is enforced by build and test procedures.
> 
>     * based on the package structure, we need to compile a coherently resolved dependency graph for all the dependencies, for each python+platform combination.
> 
>     * and then, based on this big dependency graph, we can extract the smaller sub graphs for each module / component, and use the sub graph to build the containers or run the tests.
> 
> 
> if we use bazel to compile the dependencies, this is basically what bazel will do. sadly, bazel does not have ver great support for multi-python+multi-platform, or at least it will take a lot of efforts.
> 
> that said, the dependency graph compilation for third-party dependencies is basically done by pip, which means we can do it in our own way if we want to.
> 
> maybe the challenging part is actually "properly splitting ray into sub components".

The goal of these restructurings is not to solve the dependency issues for different ray components. It is mostly about improving the state of our docker containers. We are shipping docker containers, and we ship installed dependencies in them. Users are asking for lists of these dependencies to be able to install them in their own setups, and they expect that these dependency versions are also tested - which is currently not the case.

This renaming and the follow-up (pip-compile) will ensure consistent versions between the docker images and CI. For actual test-only dependencies (in this PR the `*-test-requirements.txt` files), if we can improve the situation with bazel, that would be great!

**Comment by aslonnie** [comment]
> dependabot to make updates easier in the future.

I don't think dependabot can really help.

to update dependencies, it has to resolve dependencies. and when dependencies spreads across many different dependency files, dependabot won't be able to properly reason about the relationships among these separated files.

> It is mostly about improving the state of our docker containers.

yes, pinning down dependencies and making them explicit is good, but I think they should be generated and compiled from a bigger coherent view.

here is an example of compiled, pinned dependency list:

for release tests:

https://github.com/ray-project/ray/blob/master/release/ray_release/byod/requirements_byod_3.8.txt

for release test infra code:
https://github.com/ray-project/ray/blob/master/release/requirements_buildkite.txt

which is generated by relaxed dependency constraints like:
https://github.com/ray-project/ray/blob/master/release/requirements_buildkite.in

and in ci/tests/container compiling, we use the locked and pinned version, but we update the pinned version with the constraint, and that can be done by a bot if we have good test signals.

this workflow is totally supported by pip.

other language toolchains have similar workflows. like npm has `package.json` and `package-lock.json`:
https://github.com/ray-project/ray/blob/master/dashboard/client/package.json
https://github.com/ray-project/ray/blob/master/dashboard/client/package-lock.json

go has `go.mod` and `go.sum`:
https://go.dev/ref/mod#go-sum-files
https://github.com/golang/crypto/blob/master/go.mod
https://github.com/golang/crypto/blob/master/go.sum

I think what people want is a lock/sum file so that builds and execution results can be reproduced.

**Comment by can-anyscale** [comment]
Perhaps we can discuss more in a meeting. There is a piece about compiling these requirements into a pin-version file that I would love to understand better. I think it's important for CI and CD to share the same environment, and the compiled version sounds like a right direction, but not sure if that's what you have in mind.

**Comment by can-anyscale** [comment]
Actually book a meeting is pretty hard, but I book a pre-emptive one anyone. Let's resolve here and if not we can use that meeting. Thankkks.

**Review by richardliaw — APPROVED** [review]
given that this is mostly refactoring, I am approving this - but not in lieu of other feedback

**Comment by ericl** [comment]
At a glance, this seems to be increasing the number of different requirements files. What is the motivation for this change? It is no clear from the PR description.

Also, we shouldn't use the term "air" as it can be confusing.

**Comment by krfricke** [comment]
Apologies if this was unclear, I've updated the PR description with some more background.

In total, this PR adds 4 files: 3 `[library]-test-requirements.txt` and `ray-docker-requirements.txt`. The test requirements are split from the existing `[library]-requirements.txt` and contain packages that we don't want to ship in the `ray-ml` image, but only install in some CI jobs. Thus, splitting these off is required to fulfill this goal.

In a follow-up PR, we will use a compiled constraints file to install the same dependency versions across environments. 

I've chatted with @aslonnie and @can-anyscale and we aligned on the goal here. I'll update the PR to already include a compiled file (which will not yet be used for installation).

As for the `air` naming, I'll just rename it back to `ml` as that is the status quo and doesn't seem to have too much controversy. The reason I went with `air` originally was to include the data requirements.

**Review by aslonnie — COMMENTED** [review]
yeah, synced offline. I think the plan overall looks great. main concern is that splitting into more shards but without a story that pulls them together makes me a bit worry that it might be challenging to resolve them properly later.

so ideally, I would like to see that:

- the resolved pip-compile'd file somewhere, with all packages versions pinned
- a set of test somewhere that checks that the resolved version list satisfies every shard, ideally outside the docker build process.
- a brief doc of steps somewhere about how to add/update a dependency (how to recompile/regenerate the resolved, pinned version file)

I understand that there is going to be a follow up to have these. I am suggesting (softly) that maybe it is better to have the compiled version check in first, so that the dependencies stops wiggling before we break the requirement files into smaller pieces. :)

**Pull request merged**